index.js

```diff
@@ -1,20 +1,69 @@
-export function u(type, props, value) {
-  var node = {type: String(type)}
+/**
+ * @typedef {import('unist').Node} Node
+ * @typedef {import('unist').Parent} Parent
+ * @typedef {import('unist').Literal} Literal
+ * @typedef {Object.<string, unknown>} Props
+ * @typedef {Array.<Node>|string} ChildrenOrValue
+ *
+ * @callback BuildParentWithProps
+ * @param {string} type
+ * @param {Props} props
+ * @param {Array.<Node>} value
+ * @returns {Parent}
+ *
+ * @callback BuildParentWithoutProps
+ * @param {string} type
+ * @param {Array.<Node>} value
+ * @returns {Parent}
+ *
+ * @callback BuildLiteralWithProps
+ * @param {string} type
+ * @param {Props} props
+ * @param {string} value
+ * @returns {Literal}
+ *
+ * @callback BuildLiteralWithoutProps
+ * @param {string} type
+ * @param {string} value
+ * @returns {Literal}
+ *
+ * @callback BuildVoid
+ * @param {string} type
+ * @param {Props} [props]
+ * @returns {Node}
```

Comment:
Would it be possible to preserve some of the generic type inference from the .d.ts based typings?

Reply:
I agree this is a nice feature. As an example, the following is currently valid, but would be invalid after merging this PR as-is.

```ts
import { text } from 'mdast'

const text: Text = u('text', '')
```

Reply:
originally this was supported through overloading, but overloading isn't a part of TS through JSDoc yet https://github.com/microsoft/TypeScript/issues/25590

It might also be possible without overloads using conditional types https://www.typescriptlang.org/docs/handbook/2/conditional-types.html but not :100: sure if it would work and how well that would mix with JSDoc

Reply:
I figured out a way to solve this in JSDoc, will push `unist-util-is` with that soon and use the same idea here later

Reply:
added